Add MS_PLATFORM_CRYPTO_PROVIDER to CngProvider

DavidSimner opened the pull request:
This PR adds MS_PLATFORM_CRYPTO_PROVIDER to CngProvider, resolves #42906

The Microsoft documentation page -- https://docs.microsoft.com/en-gb/windows/win32/api/ncrypt/nf-ncrypt-ncryptopenstorageprovider#parameters -- lists 3 built-in key storage providers, but only 2 were previously present on this class.  This PR adds the third.

vcsjones commented:
Thanks! This introduces a new public API, which needs to go through the [API Review Process][1] first. Can you please [open an issue first][2] with an API Proposal?

[1]: https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md
[2]: https://github.com/dotnet/runtime/issues/new?assignees=&labels=api-suggestion&template=02_api_proposal.md&title=

DavidSimner commented:
> Thanks! This introduces a new public API, which needs to go through the [API Review Process](https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md) first. Can you please [open an issue first](https://github.com/dotnet/runtime/issues/new?assignees=&labels=api-suggestion&template=02_api_proposal.md&title=) with an API Proposal?

Done, see: https://github.com/dotnet/runtime/issues/42906

stephentoub commented:
Thanks, @DavidSimner, for the issue and the example implementation.  Per our defined processes, I'm closing this PR while the issue is discussed.  Please do feel free to re-open this or to submit a new PR if/when the suggested API is approved.

vcsjones commented:
@DavidSimner this API has been approved, so feel free to re-open this PR if you'd still like to contribute.

Regarding the test failures, this is because the reference source was not updated with the new public API. The document [`updating-ref-source.md`](https://github.com/dotnet/runtime/blob/4d784693ebc5f91c7eede32170046355ef3969b2/docs/coding-guidelines/updating-ref-source.md#for-most-assemblies-within-libraries) has steps for re-generating the reference source.

stephentoub commented on src/libraries/System.Security.Cryptography.Cng/src/System/Security/Cryptography/CngProvider.cs:
```diff
@@ -86,6 +86,14 @@ public override string ToString()
         // Well known NCrypt KSPs
         //
 
+        public static CngProvider MicrosoftPlatformCryptoProvider
```
@DavidSimner, this will need to be added to the reference assembly as well:
https://github.com/dotnet/runtime/blob/54018cf9c836c2df358ccdc6ee5e3597bd270bb6/src/libraries/System.Security.Cryptography.Cng/ref/System.Security.Cryptography.Cng.cs#L226-L227

DavidSimner replied:
Done!

carlossanlop replied:
I think the bot only adds the label when there is a modified ref file on PR creation. If a ref file is added later, the bot does not add the label. Is this correct, @maryamariyan ?

stephentoub commented on src/libraries/System.Security.Cryptography.Cng/src/System/Security/Cryptography/CngProvider.cs:
```diff
@@ -86,6 +86,14 @@ public override string ToString()
+                return s_msPlatformKsp ?? (s_msPlatformKsp = new CngProvider("Microsoft Platform Crypto Provider")); // MS_PLATFORM_CRYPTO_PROVIDER
+            }
+        }
+
```
@bartonjs, are there tests somewhere for the existing statics here?  I did a quick search but couldn't find any.

jeffhandley commented:
@DavidSimner Thanks for this contribution. Would you like to finish the PR by updating the reference source?

DavidSimner committed:
Add MS_PLATFORM_CRYPTO_PROVIDER

Taken from https://docs.microsoft.com/en-gb/windows/win32/api/ncrypt/nf-ncrypt-ncryptopenstorageprovider#parameters

DavidSimner committed:
Update reference source

DavidSimner commented:
> @DavidSimner Thanks for this contribution. Would you like to finish the PR by updating the reference source?

Done!  Please can you review and merge.  The failing builds are not related to this PR and it looks like #47380 is addressing them.

vcsjones commented on src/libraries/System.Security.Cryptography.Cng/src/System/Security/Cryptography/CngProvider.cs:
```diff
@@ -86,6 +86,14 @@ public override string ToString()
         // Well known NCrypt KSPs
         //
 
+        public static CngProvider MicrosoftPlatformCryptoProvider
```
I think there should be XML docs on new APIs. Perhaps something like this to use the other properties as a template:

```suggestion
        /// <summary>
        /// Gets a <see cref="CngProvider" /> object that specifies the Microsoft Platform Crypto Storage Provider.
        /// </summary>
        /// <value>An object that specifies the Microsoft Platform Crypto Storage Provider.</value>
        public static CngProvider MicrosoftPlatformCryptoProvider
```

@carlossanlop I noticed this PR didn't get tagged "new-api-needs-documentation". Did something change around this?

carlossanlop replied:
I think the bot only adds the label when there is a modified ref file on PR creation. If a ref file is added later, the bot does not add the label. Is this correct, @maryamariyan ?

bartonjs committed:
Update src/libraries/System.Security.Cryptography.Cng/src/System/Security/Cryptography/CngProvider.cs

Co-authored-by: Kevin Jones <vcsjones@github.com>
